File: apps/api/internal/service/payments/deposit_tracking.go
package payments

import "encoding/json"

const handledDepositTxHashesKey = "handled_deposit_tx_hashes"

func getHandledDepositTxHashes(metadata map[string]string) []string {
	if metadata == nil {
		return nil
	}
	raw, ok := metadata[handledDepositTxHashesKey]
	if !ok || raw == "" {
		return nil
	}
	var parsed []string
	if err := json.Unmarshal([]byte(raw), &parsed); err != nil {
		return nil
	}
	out := make([]string, 0, len(parsed))
	for _, v := range parsed {
		if v != "" {
			out = append(out, v)
		}
	}
	return out
}

func IsEscrowRefundTerminal(payment *Payment) bool {
	return payment.Status == "refunded" || payment.RefundTxHash != nil
}

func IsDepositTxAlreadyHandled(payment *Payment, txHash string) bool {
	if txHash == "" {
		return false
	}
	handled := getHandledDepositTxHashes(payment.Metadata)
	contains := false
	for _, h := range handled {
		if h == txHash {
			contains = true
			break
		}
	}

	if contains && payment.DepositTxHash == nil && payment.Status == "pending" {
		return true
	}
	if payment.DepositTxHash == nil || *payment.DepositTxHash != txHash {
		return false
	}
	if payment.Status == "completed" {
		return true
	}
	if IsEscrowRefundTerminal(payment) {
		return true
	}
	if payment.Status == "refunding" {
		return true
	}
	if payment.Status == "settlement_failed" && payment.RefundReason != nil {
		return true
	}
	return false
}

func MarkDepositTxHandled(metadata map[string]string, txHash string) map[string]string {
	if txHash == "" {
		if metadata == nil {
			return map[string]string{}
		}
		return metadata
	}
	out := map[string]string{}
	for k, v := range metadata {
		out[k] = v
	}
	handled := getHandledDepositTxHashes(out)
	for _, h := range handled {
		if h == txHash {
			return out
		}
	}
	handled = append(handled, txHash)
	raw, _ := json.Marshal(handled)
	out[handledDepositTxHashesKey] = string(raw)
	return out
}

File: apps/web/src/lib/payments/settlement/deposit-tracking.ts
import type { Payment } from "@/lib/db/schema";

const HANDLED_DEPOSIT_TX_HASHES_KEY = "handled_deposit_tx_hashes";

export function getHandledDepositTxHashes(
  metadata: Payment["metadata"]
): string[] {
  const raw = metadata?.[HANDLED_DEPOSIT_TX_HASHES_KEY];
  if (!raw) {
    return [];
  }

  try {
    const parsed = JSON.parse(raw) as unknown;
    if (!Array.isArray(parsed)) {
      return [];
    }

    return parsed.filter((value): value is string => typeof value === "string");
  } catch {
    return [];
  }
}

export function isDepositTxAlreadyHandled(
  payment: Payment,
  txHash: string
): boolean {
  if (!txHash) {
    return false;
  }

  const handledHashes = getHandledDepositTxHashes(payment.metadata);

  // Retry reopened after refund: metadata remembers a refunded deposit.
  if (
    handledHashes.includes(txHash) &&
    !payment.depositTxHash &&
    payment.status === "pending" &&
    payment.refundTxHash
  ) {
    return true;
  }

  if (payment.depositTxHash !== txHash) {
    return false;
  }

  // Terminal outcomes for this deposit.
  if (payment.status === "completed") {
    return true;
  }

  if (isEscrowRefundTerminal(payment)) {
    return true;
  }

  // deposit_received / refunding / settling still need processing.
  return false;
}

export function markDepositTxHandled(
  metadata: Payment["metadata"],
  txHash: string
): Record<string, string> {
  if (!txHash) {
    return metadata ?? {};
  }

  const handled = getHandledDepositTxHashes(metadata);
  if (handled.includes(txHash)) {
    return metadata ?? {};
  }

  return {
    ...(metadata ?? {}),
    [HANDLED_DEPOSIT_TX_HASHES_KEY]: JSON.stringify([...handled, txHash]),
  };
}

export function isEscrowRefundTerminal(payment: Payment): boolean {
  return payment.status === "refunded" || Boolean(payment.refundTxHash);
}
